### PR TITLE
Add Github actions

### PR DESCRIPTION
**Pull request by foikila**
Add a first take on GitHub Actions workflow for continuous integration. This workflow installs dependencies, runs unit tests, and lints the code on each push or manual trigger.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    needs: [install]
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v4
+      - name: Install dependencies
+        uses: ./.github/actions/install
+      - name: Run tests
+        run: python3 -m unittest discover
+
+  lint:
```
Comment (kh31d4r) [inline .github/workflows/main.yml:38]: if we are going to swap out pylint for ruff, maybe we could do it in a separate commit? and maybe through Make (since that is what we are using to build, still)

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    needs: [install]
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v4
+      - name: Install dependencies
+        uses: ./.github/actions/install
+      - name: Run tests
+        run: python3 -m unittest discover
```
Comment (kh31d4r) [inline .github/workflows/main.yml]: make?

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v4
+      - name: Install dependencies
+        uses: ./.github/actions/install
+
+  unit-test:
+    name: Run unit test
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
```
Comment (kh31d4r) [inline .github/workflows/main.yml]: `Ran 40 tests in 0.032s`

10 minutes seems a bit pessimistic?

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+# Ensures that only one deploy task per branch/environment will run at a time.
+concurrency:
+  group: environment-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  install:
+    name: Install dependencies
+    runs-on: ubuntu-latest
```
Comment (kh31d4r) [inline .github/workflows/main.yml]: can we have a fixed version so we know what we are running on?

File: .github/actions/install/action.yml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+runs:
+  using: composite
+  steps:
+    - name: Set up Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: '3.x'
+        cache: 'pip'
+        cache-dependency-path: .requirements.txt
+    - name: Install dependencies
```
Comment (kh31d4r) [inline .github/actions/install/action.yml:12]: make?

File: Makefile
Comment (kh31d4r) [inline Makefile:1]: these changes seem good, but unrelated to the rest

**Review by kh31d4r — COMMENTED** [review]
Maybe a more elaborate commit message?

File: .github/workflows/main.yml (hunk 2 above)
Comment (foikila) [reply .github/workflows/main.yml]: sure thing

File: .github/workflows/main.yml (hunk 3 above)
Comment (foikila) [reply .github/workflows/main.yml]: haha, yes maybe. Lowering it to 1 min

File: .github/workflows/main.yml (hunk 1 above)
Comment (foikila) [reply .github/workflows/main.yml:38]: missed that we had pylint set up already. changing to Pylint instead :+1:

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+name: Continuous integration
+
+on:
+  # Trigger the workflow on push or pull request,
+  push:
+  # Trigger the workflow via a manual run
+  workflow_dispatch:
+
+# Ensures that only one deploy task per branch/environment will run at a time.
```
Comment (AndreasArne) [inline .github/workflows/main.yml:9]: Why do we want to limit it to one task per branch?

**Review by AndreasArne — COMMENTED** [review]
Just a question
